File: vocal_flutter/lib/sheets/weight_entry_sheet.dart
// Plain weight entry — Flutter port of WeightEntrySheet.swift.
//
// Surfaces from the Progress tab so a user who just stepped on a scale
// can record the number without invoking the body-fat photo flow (which
// captures BF% and is overkill for a daily weigh-in).
//
// Persists a BodyMetric with `bodyFatPct = null` so the weight chart and
// the BF% chart stay distinct. Profile.weightLbs is updated alongside so
// the calorie-goal recomputation picks up today's number.
//
// Server-sync TODO mirrors iOS: once /api/body/weight ships, POST the
// metric here. Local-only is authoritative for the chart in the meantime.

import 'package:flutter/material.dart';
import 'package:flutter/services.dart';
import 'package:provider/provider.dart';

import '../models/models.dart';
import '../state/app_model.dart';
import '../theme/theme.dart';
import '../widgets/components.dart';

Future<void> showWeightEntrySheet(BuildContext context) {
  return showModalBottomSheet(
    context: context,
    isScrollControlled: true,
    backgroundColor: Palette.ink,
    barrierColor: Colors.black.withOpacity(0.6),
    builder: (_) => const FractionallySizedBox(
      heightFactor: 0.55,
      child: WeightEntrySheet(),
    ),
  );
}

class WeightEntrySheet extends StatefulWidget {
  const WeightEntrySheet({super.key});

  @override
  State<WeightEntrySheet> createState() => _WeightEntrySheetState();
}

enum _Units { lb, kg }

extension on _Units {
  String get label => this == _Units.lb ? 'lb' : 'kg';
}

class _WeightEntrySheetState extends State<WeightEntrySheet> {
  /// Raw user input — kept as a string so we can preserve mid-edit state
  /// across unit toggles instead of round-tripping through a double.
  final TextEditingController _controller = TextEditingController();
  final FocusNode _focus = FocusNode();
  _Units _units = _Units.lb;
  String? _validationError;
  bool _didPrefill = false;

  @override
  void initState() {
    super.initState();
    // Defer prefill until the first build so we can read AppModel via
    // context.read (init runs before the tree is fully wired).
    WidgetsBinding.instance.addPostFrameCallback((_) => _prefillIfNeeded());
  }

  @override
  void dispose() {
    _controller.dispose();
    _focus.dispose();
    super.dispose();
  }

  void _prefillIfNeeded() {
    if (_didPrefill || !mounted) return;
    _didPrefill = true;
    final app = context.read<AppModel>();
    final lb = app.bodyMetrics.isNotEmpty
        ? app.bodyMetrics.first.weightLbs
        : app.profile.weightLbs;
    if (lb > 0) {
      _controller.text = lb.toStringAsFixed(1);
    }
    _focus.requestFocus();
  }

  void _onUnitsChanged(_Units? next) {
    if (next == null || next == _units) return;
    final value = double.tryParse(_controller.text.replaceAll(',', '.'));
    final prev = _units;
    setState(() => _units = next);
    if (value == null) return;
    // Convert the buffer so the displayed magnitude stays accurate. If
    // parsing failed we leave the raw text alone — user's mid-edit.
    if (prev == _Units.lb && next == _Units.kg) {
      _controller.text = (value * 0.45359237).toStringAsFixed(1);
    } else if (prev == _Units.kg && next == _Units.lb) {
      _controller.text = (value / 0.45359237).toStringAsFixed(1);
    }
  }

  void _save() {
    final trimmed =
        _controller.text.trim().replaceAll(',', '.');
    final value = double.tryParse(trimmed);
    if (value == null) {
      setState(() => _validationError = 'Enter a number.');
      return;
    }
    // Convert to lb (on-device storage unit) before persistence. Clamp
    // to plausible adult range so a typo doesn't anchor the chart.
    final lb = _units == _Units.lb ? value : value / 0.45359237;
    if (lb <= 40 || lb >= 800) {
      setState(() =>
          _validationError = 'Out of range. Double-check the number.');
      return;
    }
    final metric = BodyMetric(
      weightLbs: lb,
      bodyFatPct: null,
      confidence: null,
      measuredAt: DateTime.now(),
    );
    final app = context.read<AppModel>();
    app.addBodyMetric(metric);
    // Keep profile aligned with most-recent measurement so calorie-goal
    // recomputation reflects today's number — otherwise the goal stays
    // anchored to a stale weight.
    app.updateProfile((p) => p.weightLbs = lb);
    HapticFeedback.mediumImpact();
    // TODO(server-sync): once /api/body/weight ships, POST `metric` here
    // so multi-device users see the entry on every install.
    Navigator.of(context).maybePop();
  }

  @override
  Widget build(BuildContext context) {
    return Container(
      color: Palette.ink,
      padding: EdgeInsets.only(
          bottom: MediaQuery.of(context).viewInsets.bottom),
      child: SafeArea(
        child: Padding(
          padding: const EdgeInsets.fromLTRB(28, 18, 28, 24),
          child: Column(
            crossAxisAlignment: CrossAxisAlignment.start,
            children: [
              _header(),
              const SizedBox(height: 18),
              _inputBlock(),
              const Spacer(),
              _saveRow(),
            ],
          ),
        ),
      ),
    );
  }

  Widget _header() {
    return Row(
      crossAxisAlignment: CrossAxisAlignment.start,
      children: [
        Expanded(
          child: Column(
            crossAxisAlignment: CrossAxisAlignment.start,
            children: [
              const Eyebrow('LOG WEIGHT', color: Palette.pulse),
              const SizedBox(height: 6),
              Text(
                'What does the scale say?',
                style: AppType.serif(28, weight: FontWeight.w500),
              ),
            ],
          ),
        ),
        GestureDetector(
          onTap: () => Navigator.of(context).maybePop(),
          child: Container(
            width: 30,
            height: 30,
            decoration: BoxDecoration(
              shape: BoxShape.circle,
              border: Border.all(color: Palette.hairlineStrong),
            ),
            child: const Icon(Icons.close, size: 13, color: Palette.ash),
          ),
        ),
      ],
    );
  }

  Widget _inputBlock() {
    return Container(
      padding: const EdgeInsets.all(20),
      decoration: cardDecoration(),
      child: Column(
        crossAxisAlignment: CrossAxisAlignment.start,
        children: [
          Row(
            crossAxisAlignment: CrossAxisAlignment.baseline,
            textBaseline: TextBaseline.alphabetic,
            children: [
              Expanded(
                child: TextField(
                  controller: _controller,
                  focusNode: _focus,
                  keyboardType: const TextInputType.numberWithOptions(
                      decimal: true),
                  style: AppType.serif(56, weight: FontWeight.w500),
                  decoration: InputDecoration(
                    isCollapsed: true,
                    border: InputBorder.none,
                    enabledBorder: InputBorder.none,
                    focusedBorder: InputBorder.none,
                    hintText: '0.0',
                    hintStyle: AppType.serif(56,
                        weight: FontWeight.w500, color: Palette.smoke),
                  ),
                ),
              ),
              const SizedBox(width: 10),
              Text(
                _units.label,
                style: AppType.serif(24,
                    weight: FontWeight.w400,
                    italic: true,
                    color: Palette.smoke),
              ),
            ],
          ),
          const SizedBox(height: 14),
          // Two-state segmented toggle. Material's SegmentedButton ships
          // with M3, picks up our colorScheme.primary (paper) cleanly.
          SegmentedButton<_Units>(
            segments: const [
              ButtonSegment(value: _Units.lb, label: Text('LB')),
              ButtonSegment(value: _Units.kg, label: Text('KG')),
            ],
            selected: {_units},
            onSelectionChanged: (s) => _onUnitsChanged(s.first),
            showSelectedIcon: false,
            style: ButtonStyle(
              textStyle: WidgetStateProperty.all(
                AppType.body(12, weight: FontWeight.w600),
              ),
            ),
          ),
          if (_validationError != null) ...[
            const SizedBox(height: 10),
            Text(
              _validationError!,
              style: AppType.body(12, color: Palette.pulse),
            ),
          ],
        ],
      ),
    );
  }

  Widget _saveRow() {
    return Row(
      children: [
        Expanded(
          child: GhostButton(
            title: 'Cancel',
            onTap: () => Navigator.of(context).maybePop(),
          ),
        ),
        const SizedBox(width: 12),
        Expanded(
          child: VoltageButton(
            title: 'Save',
            icon: Icons.check,
            onTap: _save,
          ),
        ),
      ],
    );
  }
}

File: VoCal/VoCal/WeightEntrySheet.swift
//
//  WeightEntrySheet.swift
//  VoCal
//
//  Small input sheet for logging a body-weight measurement directly from
//  the Progress tab. Weight has historically been written either by the
//  body-fat photo flow (which also captures BF%) or via HealthKit's
//  passive sync — both of those skip the case where a user just steps on
//  a scale and wants to record the number with zero camera ceremony.
//
//  Persists a `BodyMetric` with `bodyFatPct` nil so the weight chart and
//  the BF% chart stay distinct. The body-fat card on the Progress tab
//  filters by `compactMap { $0.bodyFatPct }` so a weight-only entry won't
//  pollute the BF series.
//
//  Server sync TODO: there's no plain-weight endpoint on /api today
//  (/api/bodyfat takes photos + returns BF%, not bare weight). Once a
//  `/api/body/weight` endpoint exists, swap the local-only persistence
//  block at the bottom of `save()` for a POST. The AppGroup-persisted
//  copy stays as the source of truth for the chart either way.
//

import SwiftUI
import UIKit

struct WeightEntrySheet: View {
    @EnvironmentObject private var appModel: AppModel
    @Environment(\.dismiss) private var dismiss

    /// `kg` ↔ `lb` toggle. The model always stores `weightLbs`; this is a
    /// UI-only preference that converts on save. Default reflects the
    /// most recent prior entry's apparent magnitude (anything < 100 we
    /// assume the user is thinking in kg).
    enum Units: String, CaseIterable {
        case lb, kg
        var label: String { self == .lb ? "lb" : "kg" }
    }

    @State private var raw: String
    @State private var units: Units
    @FocusState private var inputFocused: Bool
    /// Surfaced when the user types something we can't parse as a number.
    @State private var validationError: String?

    init() {
        // Pre-populate with the user's most recent weight so the typical
        // flow is "small adjustment" rather than "type from scratch".
        // Fall back to the profile weight if no history exists.
        // (We can't read AppModel from init because it's an @EnvironmentObject —
        // the @State default uses an empty string and `.onAppear` populates.)
        self._raw = State(initialValue: "")
        self._units = State(initialValue: .lb)
    }

    var body: some View {
        ZStack {
            Theme.Palette.ink.ignoresSafeArea()
            VStack(alignment: .leading, spacing: 18) {
                header
                inputBlock
                Spacer(minLength: 0)
                saveRow
            }
            .padding(.horizontal, 28)
            .padding(.top, 18)
            .padding(.bottom, 24)
        }
        .onAppear {
            // Pre-populate from the most recent body metric (or the
            // profile's static weight) so the user's typical scrub is
            // small. We avoid pre-filling on every appear (only when
            // empty) so swapping units mid-flight doesn't wipe the
            // user's pending input.
            if raw.isEmpty {
                let lb = appModel.bodyMetrics.first?.weightLbs ?? appModel.profile.weightLbs
                if lb > 0 {
                    raw = String(format: "%.1f", lb)
                    units = .lb
                }
            }
            inputFocused = true
        }
        .onChange(of: units) { old, new in
            // Convert the existing buffer when the user flips units so the
            // displayed magnitude stays accurate. If parsing fails we just
            // leave the raw text alone — the user's mid-edit.
            guard old != new, let v = Double(raw) else { return }
            switch (old, new) {
            case (.lb, .kg): raw = String(format: "%.1f", v * 0.45359237)
            case (.kg, .lb): raw = String(format: "%.1f", v / 0.45359237)
            default: break
            }
        }
    }

    private var header: some View {
        HStack(alignment: .top) {
            VStack(alignment: .leading, spacing: 6) {
                Text("LOG WEIGHT")
                    .eyebrow(Theme.Palette.pulse)
                Text("What does the scale say?")
                    .font(Theme.Font.serif(28, weight: .medium))
                    .foregroundStyle(Theme.Palette.bone)
            }
            Spacer()
            Button { dismiss() } label: {
                Image(systemName: "xmark")
                    .font(.system(size: 13, weight: .semibold))
                    .foregroundStyle(Theme.Palette.ash)
                    .frame(width: 30, height: 30)
                    .background(Circle().strokeBorder(Theme.Palette.hairlineStrong, lineWidth: 1))
            }
            .buttonStyle(.plain)
        }
    }

    private var inputBlock: some View {
        VStack(alignment: .leading, spacing: 14) {
            HStack(alignment: .firstTextBaseline, spacing: 10) {
                TextField(
                    "",
                    text: $raw,
                    prompt: Text("0.0").foregroundStyle(Theme.Palette.smoke)
                )
                .keyboardType(.decimalPad)
                .focused($inputFocused)
                .font(Theme.Font.serif(56, weight: .medium))
                .foregroundStyle(Theme.Palette.bone)
                .monospacedDigit()
                Text(units.label)
                    .font(Theme.Font.serif(24, weight: .regular, italic: true))
                    .foregroundStyle(Theme.Palette.smoke)
            }
            // Unit toggle. Picker style is `.segmented` so it reads like an
            // iOS-native control without inventing a new component for a
            // two-state choice.
            Picker("Units", selection: $units) {
                ForEach(Units.allCases, id: \.self) { u in
                    Text(u.label.uppercased()).tag(u)
                }
            }
            .pickerStyle(.segmented)
            if let err = validationError {
                Text(err)
                    .font(.system(size: 12))
                    .foregroundStyle(Theme.Palette.pulse)
            }
        }
        .padding(20)
        .background(
            RoundedRectangle(cornerRadius: Theme.Radius.md, style: .continuous)
                .fill(Theme.Palette.inkSurface)
                .overlay(
                    RoundedRectangle(cornerRadius: Theme.Radius.md, style: .continuous)
                        .strokeBorder(Theme.Palette.hairline, lineWidth: 1)
                )
        )
    }

    private var saveRow: some View {
        HStack(spacing: 12) {
            GhostButton(title: "Cancel") { dismiss() }
            VoltageButton(title: "Save", icon: "checkmark") { save() }
        }
    }

    private func save() {
        let trimmed = raw
            .trimmingCharacters(in: .whitespaces)
            .replacingOccurrences(of: ",", with: ".")
        guard let value = Double(trimmed) else {
            validationError = "Enter a number."
            return
        }
        // Convert to lb (the on-device storage unit) before persistence.
        // Clamp to a plausible adult range so a typo doesn't anchor a
        // 4.4 lb / 800 kg entry on the weight chart forever.
        let lb: Double = {
            switch units {
            case .lb: return value
            case .kg: return value / 0.45359237
            }
        }()
        guard lb > 40, lb < 800 else {
            validationError = "Out of range. Double-check the number."
            return
        }
        let metric = BodyMetric(
            weightLbs: lb,
            bodyFatPct: nil,
            confidence: nil,
            measuredAt: .now
        )
        appModel.addBodyMetric(metric)
        // Keep the profile's static weight aligned with the most-recent
        // measurement so the calorie-goal recomputation (when it next
        // runs in onboarding edit or coach refit) reflects today's
        // number — otherwise the goal stays anchored to a stale weight.
        appModel.updateProfile { p in p.weightLbs = lb }
        // Mirror to HealthKit so the Health app stays in sync. Fire-
        // and-forget; no-op if HK auth isn't granted.
        Task { await VoCalHealth.shared.write(bodyMetric: metric) }
        UIImpactFeedbackGenerator(style: .medium).impactOccurred()
        // TODO(server-sync): once /api/body/weight ships, POST `metric`
        // here so multi-device users see the entry on every install. The
        // local copy is authoritative for the chart in the meantime.
        dismiss()
    }
}

#Preview {
    WeightEntrySheet()
        .environmentObject(AppModel(
            totals: MockData.today,
            meals: [],
            profile: MockData.profile,
            bodyMetrics: MockData.bodyMetrics
        ))
        .preferredColorScheme(.dark)
}
